### PR TITLE
Added: Video/Audio statistics, replaying and corresponding examples

### PR DESCRIPTION
**Pull request by KonWol**
Implemented replaying for videos. Exemplary implementation for video statistics:
- how often did a video get replayed
- how long was each replay
- when did the video start relative to visiting the site
- how often was the video played till the end

### CODE REVIEW

File: src/questionnaireitem_media.js [hunk 1]
```diff
@@ -22,6 +22,10 @@ function QuestionnaireItemMedia(className, question, required, url, readyOnError
     this.url = url;
     this.isContentLoaded = false;
     this.stallingCount = 0;
+    this.playCount = 0;
+    this.startTime = null;
```
Comment (dennisguse) [inline src/questionnaireitem_media.js]: What is the difference to this.startTimes[] and this.startTime?
What is stored in startTimes?

File: src/questionnaireitem_media.js [hunk 2]
```diff
@@ -86,13 +90,25 @@ Start playback of playable media.
 QuestionnaireItemMedia.prototype._play = function() {
     TheFragebogen.logger.debug(this.constructor.name + "._play()", "This method must be overridden if playback is desired.");
 };
+
+QuestionnaireItemMedia.prototype._onplay = function() {
+    this.playCount += 1;
+    this.startTimes.push(new Date().getTime() - this.startTime);
+    this.setAnswer(this.getData());
```
Comment (dennisguse) [inline src/questionnaireitem_media.js]: What are you trying to achieve here? Export all stored data?
Then why implement in getData()?
Actually getData()/setData() was used to implement restore from local storage, which was later removed before open source release...

**Review by dennisguse — CHANGES_REQUESTED** [review]
At the moment it is not possible at which instances the video was actually played until the end.
Just the number of finished instances is counted.

And could `finishCount` replace `wasSccessfullyPlayed` (ie., `wasSuccessfullyPlayed` === `finishCount` > 0)?
And making the names of the attributes more explanatory would be great.

File: src/questionnaireitem_media.js [hunk 3]
```diff
@@ -86,13 +90,25 @@ Start playback of playable media.
...
+};
+
 /**
 Pause playback of playable media.
 */
 QuestionnaireItemMedia.prototype._pause = function() {
     TheFragebogen.logger.debug(this.constructor.name + "._pause()", "This method must be overridden if playback is desired.");
 };
 
+QuestionnaireItemMedia.prototype._onpause = function() {
```
Comment (dennisguse) [inline src/questionnaireitem_media.js]: Please move to a separate commit.

**Comment by KonWol** [comment]
> At the moment it is not possible at which instances the video was actually played until the end.
> Just the number of finished instances is counted.

What do you mean by that?

Removing `wasSccessfullyPlayed` is indeed possible.

**Comment by dennisguse** [comment]
@

> > At the moment it is not possible at which instances the video was actually played until the end.
> > Just the number of finished instances is counted.
> 
> What do you mean by that?

You are right: as your are setting the answer from getData(), the video duration is exported and one can determine which of the playTimes actually lead to a complete playback.

PS: I will remove `getData()` after the ES6 upgrade, as it is a left-over from far earlier times (restore state from localstorage). See #64.
Just move the code into `_onplay()'s setAnswer()` directly.

**Comment by KonWol** [comment]
I refactored most of the code and applied the same statistics also to the audio questionnaire element. Video/Audio elements now report the following results:
- An array of the resources URLs
- The total duration of the resource
- How often a stalling event occurs
- How often the user replayed the item explicitly with `replay()`. (This one does not count if the item gets shown multiple times due to pagination)
- The point in time when a replay started relative to the when the video gets shown. This allows measuring when a replay is triggered by a user.
- For each replay how often the item played.

I removed `finishCount` since it can be derived from the array length of e.g. the `videoPlayDurations` eliminating duplicates of data.

I added the additional functionality also in the `examples/example_media.js` file for testing and as a demo how to trigger video/audio replays.

**Comment by dennisguse** [comment]
Please remove the `example_media.html`.
I know it works, but it is not a "nice" way.
What do you think about implementing a` UIElementButton(className, caption, actionFunction)` where actionFunction is a callback function (can be anonymous) that is executed?
Then you can implement any action you want - for each Screen independently.
In that way, no global code is used to change internals - because that will break at some point.

Anyhow, any changes in this direction needs to be done in a separate PR.
I will take a look at the rest in the evening.

**Comment by KonWol** [comment]
> Please remove the `example_media.html`.

Done.

> What do you think about implementing a` UIElementButton(className, caption, actionFunction)` where actionFunction is a callback function (can be anonymous) that is executed?

See #66.

**Review by dennisguse — APPROVED** [review]

**Pull request merged**